### PR TITLE
Adding new method update or insert in dbmysql class

### PR DESCRIPTION
**Pull request by Badoche1337**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | not yet
| Fixed tickets | #number

This PR add a new method in dbmysql class, this method first try to perform an update on the object with specified parameters, if 0 row is modified by this update, then perform an insertOrDie on the object.

Need TU

### CODE REVIEW

File: inc/dbmysql.class.php [hunk 1]
```diff
@@ -1037,6 +1037,25 @@ function updateOrDie($table, $params, $where, $message = '') {
...
+    *
+    * @param string $table   Table name
+    * @param array  $params  Query parameters ([:field name => field value)
+    * @param array  $where   WHERE clause
+    *
+    * @return mysqli_result|boolean Query result handler
+    */
+   function updateOrInsert($table, $params, $where) {
+      $res = $this->update($table, $params, $where);
+      if ($this->affected_rows() == 0) {
```
Comment (trasher) [inline inc/dbmysql.class.php]: Are you sure `affected_rows` is not equal to 0 if record exists but ha nos been changed?

**Comment by trasher** [comment]
Ping @Badoche1337?

File: inc/dbmysql.class.php (hunk 1 above)
Comment (Badoche1337) [reply inc/dbmysql.class.php]: Indded, equal to 0 if no changes. We have to specify the CLIENT_FOUND_ROWS flag at mysqli connection to get it work (affected rows = matched rows after that). But it's a too important modification for this little feature.

I'll keep searching

Comment (trasher) [reply inc/dbmysql.class.php]: Changing mysqli connection flag is indeed a bit too muc and should impact many other features :/

Comment (trasher) [reply inc/dbmysql.class.php]: This need to be confirmed, but `update()` method apparently returns `false` when something went wrong, not if `affected_rows == 0`.
Maybe just a `if ($res !== false) {}` would do the job.

**Comment by trasher** [comment]
Stalled for 2 months. Are you going to finish this one or should we close?

File: inc/dbmysql.class.php [hunk 2]
```diff
@@ -956,6 +956,25 @@ function updateOrDie($table, $params, $where, $message = '') {
...
+    * @param array  $where   WHERE clause
+    *
+    * @return mysqli_result|boolean Query result handler
+    */
+   public function updateOrInsert($table, $params, $where) {
+      $req = $this->request($table, $where);
+      if ($req->count() == 0) {
+         return $this->insertOrDie($table, $params, 'Unable to create new element or update existing one');
+      }
+      return $this->updateOrDie($table, $params, $where, 'Unable to create new element or update existing one');
```
Comment (cedric-anne) [inline inc/dbmysql.class.php]: This is really dangerous. It should not update more thant the conflicting element. If there is multiple elements that matches the `$where` conditions, I am pretty sure that the developer does not want to update all of them.
Maybe we should only update `if ($req->count() == 1)`.

**Pull request merged**